intel_extension_for_transformers/llm/runtime/graph/models/whisper/whisper.cpp

```diff
@@ -0,0 +1,4607 @@
+      total_size += ne_nbytes(tensor);
+      model.n_loaded++;
+    }
+
+    fprintf(stderr, "%s: model size    = %7.2f MB\n", __func__, total_size / 1024.0 / 1024.0);
+
+    if (model.n_loaded == 0) {
+      fprintf(stderr, "%s: WARN no tensors loaded from model file - assuming empty model for testing\n", __func__);
+    } else if (model.n_loaded != (int)model.tensors.size()) {
+      fprintf(stderr, "%s: ERROR not all tensors loaded from model file - expected %zu, got %d\n", __func__,
+              model.tensors.size(), model.n_loaded);
+      return false;
+    }
+  }
+
+  wctx.t_load_us = model_time_us() - t_start_us;
+
+  return true;
+}
+
+// evaluate the encoder with the given state
+//
+// given audio recording (more specifically, its log mel spectrogram), runs forward pass of the encoder
+// part of the transformer model and returns the encoded features
+//
+//   - wctx:      the model
+//   - wstate:     the state of the encoder
+//   - n_threads:  number of threads to use
+//   - mel_offset: offset in the mel spectrogram (i.e. audio offset)
+//
+static bool whisper_encode_internal(whisper_context& wctx, whisper_state& wstate, const int mel_offset,
+                                    const int n_threads) {
+  const int64_t t_start_us = model_time_us();
+
+  const auto& model = wctx.model;
+  const auto& mel_inp = wstate.mel;
+  const auto& hparams = model.hparams;
+
+  const int n_ctx = wstate.exp_n_audio_ctx > 0 ? wstate.exp_n_audio_ctx : hparams.n_audio_ctx;
+  const int n_state = hparams.n_audio_state;
+  const int n_head = hparams.n_audio_head;
+  const int n_layer = hparams.n_audio_layer;
+
+  const int n_mels = hparams.n_mels;
+  assert(mel_inp.n_mel == n_mels);
+
+  struct ne_init_params params = {
+      /*.mem_size   =*/wstate.buf_compute.size(),
+      /*.mem_buffer =*/wstate.buf_compute.data(),
+      /*.no_alloc   =*/false,
+  };
+
+  struct ne_context* ctx0 = ne_init(params);
+
+  wstate.use_buf(ctx0, 0);
+
+  struct ne_tensor* mel = ne_new_tensor_2d(ctx0, NE_TYPE_F32, 2 * n_ctx, n_mels, NE_SIZE_CALC);
+  assert(mel->type == NE_TYPE_F32);
+  {
+    float* dst = (float*)mel->data;
+    memset(dst, 0, ne_nbytes(mel));
+
+    const int i0 = std::min(mel_offset, mel_inp.n_len);
+    const int i1 = std::min(mel_offset + 2 * n_ctx, mel_inp.n_len);
+
+    for (int j = 0; j < mel_inp.n_mel; ++j) {
+      for (int i = i0; i < i1; ++i) {
+        dst[j * 2 * n_ctx + (i - i0)] = mel_inp.data[j * mel_inp.n_len + i];
+      }
+    }
+  }
+
+  struct ne_tensor* cur;
+
+  {
+    // convolution + gelu
+    {
+      wstate.use_buf(ctx0, 1);
+
+      cur = ne_conv_1d_ph(ctx0, model.e_conv_1_w, mel, 1, 1);
+      cur = ne_add(ctx0, ne_repeat(ctx0, model.e_conv_1_b, cur), cur);
+
+      cur = ne_gelu(ctx0, cur);
+
+      wstate.use_buf(ctx0, 0);
+
+      cur = ne_conv_1d_ph(ctx0, model.e_conv_2_w, cur, 2, 1);
+      cur = ne_add(ctx0, ne_repeat(ctx0, model.e_conv_2_b, cur), cur);
+
+      cur = ne_gelu(ctx0, cur);
```

Comment:
can I say except the head conv nodes, wisper is just like other LLMs, applies to similar optimization (FFN, MHA, QKV)?

Reply:
Yes, it's correct.